lib/tpm2_identity_util.c

```diff
@@ -0,0 +1,480 @@
+    }
+    BIGNUM* bne = BN_new();
+    if (!bne) {
+        LOG_ERR("BN_new for bne failed\n");
+        goto error;
+    }
+    return_code = BN_set_word(bne, RSA_F4);
+    if (return_code != 1) {
+        LOG_ERR("BN_set_word failed\n");
+        BN_free(bne);
+        goto error;
+    }
+    rsa = RSA_new();
+    if (!rsa) {
+        LOG_ERR("RSA_new failed\n");
+        BN_free(bne);
+        goto error;
+    }
+    return_code = RSA_generate_key_ex(rsa, mod_size_bits, bne, NULL);
+    BN_free(bne);
+    if (return_code != 1) {
+        LOG_ERR("RSA_generate_key_ex failed\n");
+        goto error;
+    }
+    BIGNUM *n = BN_bin2bn(pub_modulus, mod_size, NULL);
+    if (n == NULL) {
+        LOG_ERR("BN_bin2bn failed\n");
+        goto error;
+    }
+    if (!RSA_set0_key(rsa, n, NULL, NULL)) {
+        LOG_ERR("RSA_set0_key failed\n");
+        BN_free(n);
+        goto error;
+    }
+    // Encrypting
+    encrypted_protection_seed->size = mod_size;
+    return_code = RSA_public_encrypt(mod_size, encoded,
+            encrypted_protection_seed->secret, rsa, RSA_NO_PADDING);
+    if (return_code < 0) {
+        LOG_ERR("Failed RSA_public_encrypt\n");
+        goto error;
+    }
+
+    rval = true;
+
+error:
+    free(pub_modulus);
+    RSA_free(rsa);
+    return rval;
+}
+
+bool tpm2_identity_util_calc_outer_integrity_hmac_key_and_dupsensitive_enc_key(
+        TPM2B_PUBLIC *parent_pub,
+        TPM2B_NAME *pubname,
+        TPM2B_DIGEST *protection_seed,
+        TPM2B_MAX_BUFFER *protection_hmac_key,
+        TPM2B_MAX_BUFFER *protection_enc_key) {
+
+    TPM2B null_2b = { .size = 0 };
+
+    TPMI_ALG_HASH parent_alg = parent_pub->publicArea.nameAlg;
+    UINT16 parent_hash_size = tpm2_alg_util_get_hash_size(parent_alg);
+
+    TSS2_RC rval = tpm2_kdfa(parent_alg, (TPM2B *)protection_seed, "INTEGRITY",
+            &null_2b, &null_2b, parent_hash_size * 8, protection_hmac_key);
+    if (rval != TPM2_RC_SUCCESS) {
+        return false;
+    }
+
+    TPM2_KEY_BITS pub_key_bits = get_pub_asym_key_bits(parent_pub);
+
+    rval = tpm2_kdfa(parent_alg, (TPM2B *)protection_seed, "STORAGE",
+            (TPM2B *)pubname, &null_2b, pub_key_bits,
+            protection_enc_key);
+    if (rval != TPM2_RC_SUCCESS) {
+        return false;
+    }
+
+    return true;
+}
+
+
+bool tpm2_identity_util_encrypt_seed_with_public_key(TPM2B_DIGEST *protection_seed,
+        TPM2B_PUBLIC *parent_pub, unsigned char *label, int labelLen,
+        TPM2B_ENCRYPTED_SECRET *encrypted_protection_seed) {
+    bool result = false;
+    TPMI_ALG_PUBLIC alg = parent_pub->publicArea.type;
+    
+    switch (alg) {
+    case TPM2_ALG_RSA:
+        result = encrypt_seed_with_tpm2_rsa_public_key(protection_seed, 
+                parent_pub, label, labelLen, encrypted_protection_seed);
+        break;
+    case TPM2_ALG_ECC:
+        LOG_ERR("Algorithm '%s' not supported yet", tpm2_alg_util_algtostr(alg,
+            tpm2_alg_util_flags_any));
+        result = false;
```

Comment:
Do you have plans to support ECC in the short term? Just out of curiosity.

Reply:
We are definitely interested in adding ECC support, though we don't have an exact timeline on when this would be implemented.  At least for now the priority has been for having RSA support, but ECC support is also a feature we are interested in having for keylime (see https://github.com/keylime/python-keylime/issues/28).